java/core/src/main/java/co/worklytics/psoxy/storage/impl/ColumnarBulkDataSanitizerImpl.java

```diff
@@ -162,94 +184,146 @@ public void sanitize(@NonNull Reader reader,
+            columnsToRename.entrySet().stream().filter(e -> e.getValue().equalsIgnoreCase(a)).findFirst().map(e -> records.getHeaderMap().getOrDefault(e.getKey(), Integer.MAX_VALUE)).orElse(Integer.MAX_VALUE)));
+        Comparator<String> byColumnName = Comparator.naturalOrder();
+
+        List<String> columnNamesForOutputFile = transformTable.rowKeySet()
+            .stream()
+            .sorted(originalHeadersOrRenamedFirst.thenComparing(byColumnName))
+            .collect(Collectors.toList())
+            .stream()
+            // leave original casing for headers
+            .map( h -> headers.stream().filter( o -> h.equalsIgnoreCase(o)).findFirst().orElse(h))
+            .collect(Collectors.toList());
+
         CSVFormat csvFormat = CSVFormat.Builder.create()
             .setHeader(columnNamesForOutputFile.toArray(new String[0]))
             .build();
 
+        // immutable map uses insertion order
+        // create an empty record to fill with the transformed values, ensuring all rows have
+        // the same columns
+        ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+        columnNamesForOutputFile.forEach(h -> builder.put(h, ""));
+        ImmutableMap<String,String> emptyRow = builder.build();
+
         try (CSVPrinter printer = new CSVPrinter(writer, csvFormat)) {
             UnmodifiableIterator<List<CSVRecord>> chunks =
                 Iterators.partition(records.iterator(), this.getRecordShuffleChunkSize());
 
             for (UnmodifiableIterator<List<CSVRecord>> chunkIterator = chunks; chunkIterator.hasNext(); ) {
                 List<CSVRecord> chunk = new ArrayList<>(chunkIterator.next());
-                shuffleImplementation.apply(chunk).forEach(row -> {
-                    Stream<Object> sanitized =
-                        headers.stream() // only iterate on allowed columns (redaction implicit by omission)
-                            .map(column -> Pair.of(columnsToRename.getOrDefault(column, column).toLowerCase(), row.get(column)))
-                            .map(p -> applyPseudonymizationIfAny.apply(p.getKey(), p.getValue()));
-
-
-                    //add duplicated columns, transformed/pseudonymized as necessary
-                    sanitized = Streams.concat(sanitized, columnsToDuplicate.entrySet().stream()
-                        .map(entry -> {
-                            if (columnsToTransform.containsKey(entry.getKey().toLowerCase())) {
-                                return applyTransformIfAny.apply(entry.getKey(), row.get(entry.getKey()));
-                            } else {
-                                return applyPseudonymizationIfAny.apply(entry.getValue(), row.get(entry.getKey()));
-                            }
-                        }));
 
+                shuffleImplementation.apply(chunk).forEach(record -> {
+
+                    LinkedHashMap<String, String> newRecord = new LinkedHashMap<>(emptyRow);
```

Comment:
As a suggestion, moving this outside of the foreach and done a `newRecord.values().clean()` after print could save some memory